include/deal.II/lac/utilities.h

```diff
@@ -0,0 +1,160 @@
+                                      VectorMemory<VectorType> &vector_memory);
+  }
+
+}
+
+
+/*------------------------- Implementation ----------------------------*/
+
+#ifndef DOXYGEN
+
+namespace Utilities
+{
+  namespace Lac
+  {
+    template <typename OperatorType, typename VectorType>
+    double Lanczos_largest_eigenvalue(const OperatorType &H,
+                                      const VectorType &v0_,
+                                      const unsigned int k,
+                                      VectorMemory<VectorType> &vector_memory)
+    {
+      // Do k-step Lanczos:
+
+      typename VectorMemory<VectorType>::Pointer v(vector_memory);
+      typename VectorMemory<VectorType>::Pointer v0(vector_memory);
+      typename VectorMemory<VectorType>::Pointer f(vector_memory);
+
+      v->reinit(v0_);
+      v0->reinit(v0_);
+      f->reinit(v0_);
+
+      // two vectors to store diagonal and subdiagonal of the Lanczos
+      // matrix
+      std::vector<double> diagonal;
+      std::vector<double> subdiagonal;
+
+      // scalars to store norms and inner products
+      double a = 0, b = 0;
+
+      // 1. Normalize input vector
+      (*v) = v0_;
+      a = v->l2_norm();
+      Assert (a!=0, ExcDivideByZero());
+      (*v) *= 1./a;
+
+      // 2. Compute f = Hv; a = f*v; f <- f - av; T(0,0)=a;
+      H.vmult(*f,*v);
+      a = (*f)*(*v);
+      f->add(-a,*v);
+      diagonal.push_back(a);
+
+      // 3. Loop over steps
+      for (unsigned int i = 1; i < k; ++i)
+        {
+          // 4. L2 norm of f
+          b = f->l2_norm();
+          Assert (b!=0, ExcDivideByZero());
+          // 5. v0 <- v; v <- f/b
+          *v0 = *v;
+          *v = *f;
+          (*v) *= 1./b;
+          // 6. f = Hv; f <- f - b v0;
+          H.vmult(*f,*v);
+          f->add(-b, *v0);
+          // 7. a = f*v; f <- f - a v;
+          a = (*f) * (*v);
+          f->add(-a, *v);
+          // 8. T(i,i-1) = T(i-1,i) = b;  T(i,i) = a;
+          diagonal.push_back(a);
+          subdiagonal.push_back(b);
+        }
+
+      Assert (diagonal.size() == k,
+              ExcInternalError());
+      Assert (subdiagonal.size() == k-1,
+              ExcInternalError());
+
+      // Use Lapack dstev to get ||T||_2 norm, i.e. the largest eigenvalue
+      // of T
+      const int n = k;
+      std::vector<double> Z;    // unused for eigenvalues-only ("N") job
+      const int ldz = 1;        // ^^   (>=1)
+      std::vector<double> work; // ^^
+      int info;
+      // call lapack_templates.h wrapper:
+      stev ("N", &n,
+            &diagonal[0], &subdiagonal[0],
+            &Z[0], &ldz, &work[0],
+            &info);
+
+      Assert (info == 0,
+              LAPACKSupport::ExcErrorCode("dstev", info));
+
+      // note that the largest eigenvalue of T is below the largest
+      // eigenvalue of the operator.
+      // return ||T||_2 + ||f||_2, although it is not guaranteed to be an upper bound.
```

Comment:
oh, I think it would be worth pointing out in the documentation that the upper bound is then only an estimate, not really a hard upper bound!